### ISSUE TITLE
Problem building crate on sqlx dependency

### ISSUE THREAD

**Issue by alerque** [novel-rs/cli#92]
I've been trying to setup proper Arch Linux packaging for this and am having trouble building it in a clean chroot. I get stuck on this error:

```
<truncated>
   Compiling opencc-rs v0.3.0
   Compiling sea-schema v0.11.0
   Compiling trust-dns-resolver v0.22.0
   Compiling hyper v0.14.26
   Compiling sqlx v0.6.3
error: /build/novel-cli/src/cli-0.3.1/target/release/deps/libsqlx_macros-a6d2a8db0bd21b7c.so: undefined symbol: sqlite3_unlock_notify
  --> /build/.cargo/registry/src/github.com-1ecc6299db9ec823/sqlx-0.6.3/src/lib.rs:77:9
   |
77 | pub use sqlx_macros::test;
   |         ^^^^^^^^^^^

error: could not compile `sqlx` due to previous error
warning: build failed, waiting for other jobs to finish...
```

This is with Rust 1.69.0, and the chroot I'm working in has `clang` and `cmake` available in addition to the stuff in Arch's base-devel group that covers a lot of generic build tooling. As far as I can tell from both this project and *sqlx* this should be sufficient. I've messed around with feature flags, using or not using `--frozen`, having `sqlite` available in the chroot, and more. So far I can't figure out what is causing this. Any ideas?

**Comment by TerakomariGandesblood** [comment]
Unfortunately, I was not able to reproduce this issue, I tried building on Ubuntu and WSL2 with no issues.
According to the error message, it may be a problem with sqlx? I'm not sure.
Also, the Minimal Supported Rust Version of this project is 1.70.
@alerque

**Comment by alerque** [comment]
That means you released 7 hours after your upstream minimum supported Rust version. Definitely should be documented! By the way you can codify it like this in `Cargo.toml`:

```toml
[package]
rust = "1.70"
```

Documentation: https://rust-lang.github.io/rfcs/2495-min-rust-version.html

**Comment by alerque** [comment]
I just confirmed this is NOT a result of the Rust version. Using 1.70.0 in the chroot fails with the same error. Some other dependency that exists on my personal desktop is missing from the chroot.

**Comment by TerakomariGandesblood** [comment]
Someone seems to have the same problem. [rust#111691](https://github.com/rust-lang/rust/issues/111691). I don't quite understand what caused this.

**Comment by alerque** [comment]
I think my problem is actually many different problems. I too am most confused about the rust crate resolution one, but honestly *all* the error states are kind of confusing.



1. My initial error message in the OP here can be replicated by building without `sqlite` v3 header files available on the system AND with Rust 1.69 or Rust 1.70. The confusing bit is that the error *stays* exactly like that even when having sqlite3 headers available when building with Rust 1.69, but changes to the later crate resolution error when building with sqlite3 available and Rust 1.70.

1. That leaves us using Rust 1.70, but a weird error about resolving crates we don't understand. It's completely reproducible for me, but while chasing building another package I edited the wrong one to disable LTO flags and quite unexpectedly that made a difference. Setting `export CFLAGS+=" -ffat-lto-objects"` before calling `cargo build` avoids the problems building sqlx. Now the build gets all the way

	```
	<truncated>
	Compiling warp v0.3.5
	Compiling reqwest v0.11.18
	Compiling reqwest_cookie_store v0.5.0
	Compiling boring v2.1.0
	Compiling sqlx-core v0.6.3
	Compiling sqlx-macros v0.6.3
	Compiling sqlx v0.6.3
	Compiling sea-query-binder v0.3.1
	Compiling sea-orm v0.11.3
	Compiling sea-orm-migration v0.11.3
	Compiling novel-api v0.7.1
	Compiling novel-cli v0.3.1 (/build/novel-cli/src/cli-0.3.1)











	^Nerror: linking with `cc` failed: exit status: 1
	|
	= note: LC_ALL="C" PATH="/usr/lib64/rustlib/x86_64-unknown-linux-gnu/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl" VSLANG="1033" "cc" "-m64" "/tmp/rustcicbe2Y/symbols.o" "/build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/deps" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/libwebp-sys-84d7ab54ca1a0019/out" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/zstd-sys-c718aa1ca3e49cf0/out" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/boring-sys-e239db7c579ab80a/out/build/" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/ring-338f65b2264a8bae/out" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/libsqlite3-sys-72fe6d40bcd938f9/out" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/opencc-sys-7a876203c9291b77/out/build/deps/marisa-0.2.6" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/opencc-sys-7a876203c9291b77/out/build/src" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/link-cplusplus-63557613b046e445/out" "-L" "/build/novel-cli/src/cli-0.3.1/target/release/build/snmalloc-sys-4cd043afedba46b7/out/./build" "-L" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/rustcicbe2Y/libopencc_sys-354dcc76370de7ad.rlib" "/tmp/rustcicbe2Y/liblink_cplusplus-3514f0a9a08c4856.rlib" "/tmp/rustcicbe2Y/liblibsqlite3_sys-b936ea7909f711ce.rlib" "/tmp/rustcicbe2Y/liblibwebp_sys-ed84d4292940e7da.rlib" "/tmp/rustcicbe2Y/libzstd_sys-7ec1c9e467022b50.rlib" "/tmp/rustcicbe2Y/libring-e3fec540d4a81568.rlib" "/tmp/rustcicbe2Y/libboring_sys-bf0784375ffd9f7b.rlib" "/usr/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-ae3a35d29261b98a.rlib" "-Wl,-Bdynamic" "-lstdc++" "-lsnmallocshim-rust" "-latomic" "-lc_nonshared" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/usr/lib64/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs"
	= note: /usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `<opencc_rs::OpenCC as core::ops::drop::Drop>::drop':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:90: undefined reference to `opencc_close'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:90: undefined reference to `opencc_close'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:90: undefined reference to `opencc_close'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::new':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:37: undefined reference to `opencc_open'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::convert':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:(.text._ZN9opencc_rs6OpenCC7convert17h50c0ad8f75361f45E+0x1de): undefined reference to `opencc_convert_utf8_free'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::convert':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:(.text._ZN9opencc_rs6OpenCC7convert17h9f85560c28ec4eddE+0x1de): undefined reference to `opencc_convert_utf8_free'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::convert':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:(.text._ZN9opencc_rs6OpenCC7convert17hfada1c2abcfe7dd5E+0x1de): undefined reference to `opencc_convert_utf8_free'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `boring_sys::init::{{closure}}':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/boring-sys-2.1.0/src/lib.rs:78: undefined reference to `OPENSSL_init_ssl'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::convert':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:(.text._ZN9opencc_rs6OpenCC7convert17h74375e7067509576E+0x1de): undefined reference to `opencc_convert_utf8_free'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `<opencc_rs::OpenCC as core::ops::drop::Drop>::drop':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:90: undefined reference to `opencc_close'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::new':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:37: undefined reference to `opencc_open'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `<opencc_rs::OpenCC as core::ops::drop::Drop>::drop':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:90: undefined reference to `opencc_close'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::convert':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:(.text._ZN9opencc_rs6OpenCC7convert17h7bec18dfde13fc8bE+0x299): undefined reference to `opencc_convert_utf8_free'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `opencc_rs::OpenCC::convert':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:62: undefined reference to `opencc_convert_utf8'
			/usr/bin/ld: /build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:(.text._ZN9opencc_rs6OpenCC7convert17hf948063c01dad928E+0x1de): undefined reference to `opencc_convert_utf8_free'
			/usr/bin/ld: /build/novel-cli/src/cli-0.3.1/target/release/deps/novel_cli-f70b876d53580709.novel_cli.2efa0bf1-cgu.0.rcgu.o: in function `<opencc_rs::OpenCC as core::ops::drop::Drop>::drop':
			/build/.cargo/registry/src/index.crates.io-6f17d22bba15001f/opencc-rs-0.3.0/src/lib.rs:90: undefined reference to `opencc_close'
			collect2: error: ld returned 1 exit status

	= note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
	= note: use the `-l` flag to specify native libraries to link
	= note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

	error: could not compile `novel-cli` (bin "novel-cli") due to previous error
	```

	In other words building sqlx has issues with normal distro provided LTO related CFLAGS. Why this changes/avoids the error resolving crates I don't know, perhaps just having the different flags enabled is changing what features are enabled inside a library crate somewhere?

What that still doesn't get me is a working build method to build in a clean chroot. On my host system with Rust 1.70.0 I can get a build now, but even *with* sqlite headers installed in the chroot we seem to have our choice of two problems: crate resolution issues or linker issues.

I still suspect the answer is still some *missing* dependency that needs to be installed in order to build correctly, but so far I can't isolate what that is.

**Comment by TerakomariGandesblood** [comment]
Dependencies have changed a lot now and this question may no longer be valid.
